missing: add quiet and json

ilovezfs opened the pull request:
- default is verbose, unless the output is going to a pipe, in which
  case the default is quiet (cf. `outdated`)
- verbose is almost the same as the pre-existing `missing` output, with
  the only change being that both the dependent and dependency are
  printed as `full_name` instead of the current behavior, which prints
  the dependent as `full_name` and the dependency as `name`
- quiet outputs a single, non-repeating list of missing formulae, each
  printed as its `full_name`
- json outputs the same information as verbose, one entry per dependent
  with a missing dependency, each entry containing the dependent itself
  associated with the key "name" and its dependencies in an array mapped
  to the key "missing"

ilovezfs commented:
@mikemcquaid this probably needs the maintainer feedback tag.

ilovezfs commented on Library/Homebrew/cmd/missing.rb:
```diff
@@ -6,15 +6,40 @@ module Homebrew
+        puts "#{m * " "}"
+      end
+    end
+    s.each { |f| puts f.full_name } unless verbose
```
or is `puts_columns s.map(&:full_name)` preferred?

ilovezfs replied:
also, should the set be sorted in any particular way, alphabetical order, dependency order with leaves last, or something else?

UniqMartin replied:
I'm in favor of `puts_columns` and sorted for sure. Order should be either simply alphabetical or core formulae followed by formulae from taps (see `brew list --full-name`; I'm afraid we're not very consistent with that across various commands).

mikemcquaid commented:
Not obvious from the code: is the non-JSON non-quiet output identical?

ilovezfs commented:
Yup, unless it's going through a pipe, in which case it defaults to quiet, which I suspect is the cause of the test failure. Same behavior (and nearly identical code) to what's in the `outdated` command.

ilovezfs commented:
Test failure was caused by the exit status, so I made it zero like the outdated command unless formulae have been specified, and re-pushed.

ilovezfs commented on Library/Homebrew/cmd/missing.rb:
```diff
@@ -6,15 +6,40 @@ module Homebrew
+    else
+      missing = print_missing(formulae)
+    end
+    Homebrew.failed = ARGV.resolved_formulae.any? && missing.any?
```
It may be appropriate to actually have a non-zero exit status if any are missing regardless of whether formulae have been specified, but that would require changing the test.

UniqMartin replied:
I'm in favor of making the exit status independent of whether formulae have been provided as arguments or not. If that entails changing a test, so be it.

ilovezfs replied:
Again, this is the same logic and return value calculation as `outdated`. Do you want both `missing` and `outdated` returning 0 or returning 1 when there's something found?

UniqMartin replied:
Yes, in my opinion that should be changed, too. (Obviously in a separate PR.)

ilovezfs replied:
Oh, OK, I was thinking I'd just rename this one, but if you'd like them separated that's fine.

ilovezfs replied:
@UniqMartin To be clear you want `missing` and `outdated` to return 0 for empty output (none missing/outdated) and 1 for non-empty output (one or more formulae missing/outdated), correct?

UniqMartin replied:
Correct. Or to phrase it slightly differently, it makes sense to me that the exit status for `brew missing` and `brew missing $(brew list)` is consistent.

ilovezfs commented:
OK, looks like the tests now passed. Re-pushed with the man page updated, as well.

ilovezfs commented:
@mikemcquaid OK, I updated the PR just now in a way that does make a change to the existing output, namely that both the formulae that have something and the missing formulae will be printed as their full name. Currently, without this PR, only the formulae that have something missing are printed as full name, but the missing formulae themselves are printed just as the simple name.

old:
```
homebrew/dupes/nano: ncurses
```

new:
```
homebrew/dupes/nano: homebrew/dupes/ncurses
```

UniqMartin commented on Library/Homebrew/cmd/missing.rb:
```diff
@@ -6,15 +6,41 @@ module Homebrew
+    if ARGV.json == "v1"
+      missing = print_missing_json(formulae)
+    else
+      missing = print_missing(formulae)
```
In Ruby it is preferable to assign the return value of the `if` expression instead of having separate assignments in its various branches (see the `if ARGV.named.empty?` directly above).

ilovezfs replied:
@UniqMartin Sure, but I guess we'll be changing this, too, then? https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/outdated.rb#L7-L10
;)

UniqMartin replied:
Indeed (another thing that could be added to a PR for `outdated`).

These kinds of issues should in theory be flagged by `brew style <filename>`, which gives good advise most of the time and should be taken into account for new code. Unfortunately, running it on all of Homebrew is currently not very helpful, because we've never did a mass cleanup of our code after `brew style` was introduced.

UniqMartin replied:
On second thought, maybe it shouldn't be changed in `outdated` because that would be a purely stylistic change. Here, on the other hand, we're talking about new code and that should be written with the currently preferred style in mind.

ilovezfs replied:
The code in the two commands with this PR is nearly identical. I almost wonder if they should be refactored into a single command under the hood. But either way, it would be a shame to have them diverge, I think.

UniqMartin replied:
Personally, I would approve of a separate commit (as part of the PR that changes `outdated`) that would clean up the style therein. I'm not sure other maintainers agree with me on that.

Either way, the commands are dealing with sufficiently different things that their code should mostly remain separated. If refactoring some common code into a shared utility function or something like that makes sense, that could be done. (But I currently don't see enough overlap between the two and where that would be the case.)

UniqMartin commented on Library/Homebrew/cmd/missing.rb:
```diff
@@ -6,15 +6,41 @@ module Homebrew
+        m.each { |f| s << f }
+      end
+    end
+    s.each { |f| puts f.full_name } unless verbose
```
I find the intermixed parts that deal with verbose and non-verbose output a bit hard to read. I don't think it is necessary to populate a new set when the same information is also duplicated in the `missing` hash. It just needs to be slightly transformed for the non-verbose output.

UniqMartin commented on Library/Homebrew/cmd/missing.rb:
```diff
@@ -6,15 +6,41 @@ module Homebrew
+    json = []
+    missing = Diagnostic.missing_deps(formulae) do |name, m|
+      json << { :name => name,
+                :missing => m.collect(&:full_name) }
```
A bit like in a comment further up, I think it would be more idiomatic Ruby (and easier to read), if `json` was created by doing something like the following (after invoking `Diagnostic.missing_deps` without a block).

```ruby
json = missing.map do |formula_name, missing_deps|
  { :name => formula_name, :missing => missing.deps.map(&:full_name) }
end
```

ilovezfs replied:
And we'll be changing this as well, then, I guess
https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/outdated.rb#L27-L37

UniqMartin replied:
As remarked in a different comment, it probably shouldn't be changed in `outdated` because that would be a purely stylistic change. Here, on the other hand, we're talking about new code and that should be written with the currently preferred style in mind.

UniqMartin commented:
Overall, nice improvements to this command, making it a lot more useful. :+1:

> Re-pushed with the man page updated, as well.

Please make those edits in [`brew.1.md`](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/manpages/brew.1.md) and then run `brew man` to regenerate the man page and its HTML version, making sure to include all three files in the commit.

> OK, I updated the PR just now in a way that does make a change to the existing output, namely that both the formulae that have something missing and the missing formulae will be printed as their full name.

That's a very good change. Makes much more sense to refer to formulae consistently.

ilovezfs commented:
>Overall, nice improvements to this command, making it a lot more useful. :+1:

Thanks. I was quite surprised `brew install $(brew missing)` or `brew missing | ...` would require extra parsing, which was the reason for looking into this.


>That's a very good change. Makes much more sense to refer to formulae consistently.

The old code was using name not full_name because it was relying on the to_s conversion for formulae:
https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/formula.rb#L1034-L1036

I wonder if that should be changed to full_name, but I'm not sure how widely the automatic conversion to just simple name is used in the code base.

UniqMartin commented:
> The old code was using name not full_name because it was relying on the to_s conversion for formulae:
> https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/formula.rb#L1034-L1036
> 
> I wonder if that should be changed to full_name, but I'm not sure how widely the automatic conversion to just simple name is used in the code base.

In general, this sounds like a very reasonable suggestion to me, but given the wide impact of such a change it would require careful testing and auditing of the existing code. That's definitely something for a separate PR and also requires feedback from more seasoned maintainers that me. :smile_cat:

ilovezfs commented:
@mikemcquaid @UniqMartin I updated the PR with the changes Martin proposed.

Also,
1. I factored out the common piece of the `missing` and `outdated` commands into a separate `FormulaSubsetPrinter` class (or whatever you'd like it called).
2. Made the same "style" updates to outdated.
3. Made the same exit code modification to outdated: 0 if none outdated, 1 otherwise.
4. Updated the tests accordingly.

ilovezfs commented on Library/Homebrew/formula_subset_printer.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+
+    def quiet
+      @subset.class == Hash ? @subset.values.flatten.uniq : @subset
```
Alternatively, it could be either
- abstract and always implemented in subclass

or

- defined here as `@subset` by default, and then overridden in `missing` as `@subset.values.flatten.uniq`

ilovezfs commented on Library/Homebrew/formula_printer.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+
+    def quiet
+      @selection.class == Hash ? @selection.values.flatten.uniq : @selection
```
Alternatively, it could be either
- abstract and always implemented in subclass

or

- defined here as `@subset` by default, and then overridden in `missing` as `@subset.values.flatten.uniq`

ilovezfs commented:
@xu-cheng Any opinions/thoughts/suggestions on the FormulaPrinter class and the PR?

mikemcquaid commented:
> Personally, I found FormulaPrinter make the code harder to read than current condition. But let's see what's the other maintainer's opinion.

I agree. I think it'd be nicer to just have some repetition for now unless we can find 3+ commands which FormulaPrinter. The name is generic enough that it's hard to know exactly what it's doing.

ilovezfs commented:
@mikemcquaid I think `brew leaves` can use it too.

UniqMartin commented:
I think `FormulaPrinter` is currently doing too much, particularly because the output of `brew missing` and `brew outdated` is not as similar as it seems:

- `brew missing`:
  - Verbose: `<formula>: <missing dependents>`
  - Quiet/non-TTY: Just the right-hand side of the verbose output, and de-duplicated.
- `brew outdated`:
  - Verbose: `<formula>: <version information>`
  - Quiet/non-TTY: Just the left-hand side of the verbose output.

Nevertheless, I do like that `FormulaPrinter` simplifies the logic for deciding whether to output the JSON, quiet, or verbose version of the data. It would be great if this aspect could be refactored so that the already adapted (and possibly other) commands can use this. (It will ultimately make those command react more consistently to various combinations of arguments and the presence/absence of a TTY.)

ilovezfs commented:
@UniqMartin yes, that's why I asked https://github.com/Homebrew/legacy-homebrew/pull/50273/files#r56972013

The heuristic I'm suggesting is that if the selection is a hash, it's reasonable to assume we want the values for the quiet output, whereas if the selection is a list, I think it's reasonable to assume we want the list for the quiet output. I don't see it as left hand vs. right hand. Is there a counterexample where the selection would be a hash but we don't want values as the quiet output? Is there a counterexample where the selection would be a list and we don't want the list as the quiet output?

And of course the methods can be overridden in the subclass so it's not preventing you from doing something else if you needed to.

UniqMartin commented:
Yes, the outlined heuristic hash vs. list happens to work out. I'm not arguing with that, just claiming that this hidden interpretation of those data structures is probably too obscure, when hidden deep inside `FormulaPrinter`. I can't come up with an example where that wouldn't work, but I also have a hard time coming up with any somewhat similar commands besides `leaves` (that you already mentioned).

ilovezfs commented:
Another approach might be to make outdated, missing, and leaves aliases for `brew list --outdated`, `brew list --missing`, and `brew list --leaves`, get rid of those three commands, and handle them in the same way `brew list --pinned` is handled, or conversely split `brew list --pinned` out as a separate command `brew pinned` that used formula printer.

MikeMcQuaid commented:
Can this be ported over to the Homebrew/brew repo when it's ready to 🚢? Thanks!
